packages/starlight/utils/plugins.ts

```diff
@@ -0,0 +1,143 @@
+import type { AstroIntegration, AstroIntegrationLogger } from 'astro';
+import { z } from 'astro/zod';
+import type { StarlightUserConfig } from '../utils/user-config';
+import { errorMap } from '../utils/error-map';
+
+/**
+ * Runs Starlight plugins in the order that they are configured after validating the plugins config
+ * and returns the final user config that may have been updated by the plugins and a list of any
+ * integrations added by the plugins.
+ * Note that the returned config is not validated and should still be validated by the caller.
+ */
+export async function runPlugins(
+	{ plugins, ...config }: StarlightUserConfigWithPlugins,
+	logger: AstroIntegrationLogger
+) {
+	const parsedPluginsConfig = starlightPluginsConfigSchema.safeParse({ plugins }, { errorMap });
+
+	if (!parsedPluginsConfig.success) {
+		throw new Error(
+			'Invalid plugins config passed to starlight integration\n' +
+				parsedPluginsConfig.error.issues.map((i) => i.message).join('\n')
+		);
+	}
+
+	let userConfig = config;
+	const integrations: AstroIntegration[] = [];
+
+	for (const { name, plugin } of parsedPluginsConfig.data.plugins) {
```

Comment:
Should we support processing plugins added by preceding plugins? If so, maybe we need something like this?

```suggestion
	for (let i = 0; i < userConfig.plugins.length; i++) {
		const { name, plugin } = userConfig.plugins[i];
```

Reply:
I thought about this during the initial prototype and decided to leave it out for now. My reasoning (and maybe wrong) was that it would maybe be more "natural"/explicit to add a `addPlugin` like `addIntegration` considering that we only want to support adding plugins but not remove/re-order them while running them which I think could be less obvious with a helper called `updateConfig`.

If we decide to allow this in `updateConfig`, we would indeed need to tweak the syntax like you suggested and also extract the updated plugins config from the updated Starlight config and also re-run the plugin validation obviously.

Reply:
That does make sense to me.

I will share some thoughts from the discussion of when we enabled integrations to add integrations in Astro’s `updateConfig()`. Initially I made an `addIntegration()` helper, but Blu pointed out that ideally `updateConfig()` works for all config options so a user doesn’t have to know what they can and can’t use that hook for (otherwise you risk proliferating `updateSidebar()`, `updateHead()` etc. for any more complex option).

I think the issue in Starlight’s case is that Astro’s [`mergeConfig()`](https://github.com/withastro/astro/blob/262cef2487c7494bd8d23b4ab27bfcdf1870a111/packages/astro/src/core/config/merge.ts) is generally able to not worry about order, so merging `{ option: [something] }` and `updateConfig({ option: [somethingElse] })` as `{ option: [something, somethingElse] }` is safe (and also nicer for the dev to not worry about spreading all the right things). But in Starlight I can totally imagine wanting to do stuff like prepend or filter sidebar items etc., which the `mergeConfig()` approach prevents. So I think I agree with forcing plugin developers to merge config manually even if this does place some trust in them to not break user config.

`plugins` is the only problematic field here right? I wonder if we should error if a plugin tries to mutate it?

Reply:
> … Blu pointed out that ideally `updateConfig()` works for all config options so a user doesn’t have to know what they can and can’t use that hook for …

This is a very good point. 

> I think the issue in Starlight’s case is that Astro’s `mergeConfig()` is generally able to not worry about order

My initial implementation actually had a `mergeConfig()` which was a slimmed down version of Astro’s `mergeConfig()` but as you mentioned, this is not ideal in the case a Starlight.

> `plugins` is the only problematic field here right? 

This is indeed the only one.

> I wonder if we should error if a plugin tries to mutate it?

This would already happen when validating the config after the plugins have run but maybe you mean adding an explicit check in the `updateConfig()` hook?

Reply:
> This would already happen when validating the config after the plugins have run but maybe you mean adding an explicit check in the `updateConfig()` hook?

I think you answered this in https://github.com/withastro/starlight/pull/942#discussion_r1378878069 but validating in `updateConfig()` or after would both be fine. I guess inside `updateConfig()` we might have the advantage of a nice error message that says which plugin was responsible.